File: 1552. Magnetic Force Between Two Balls.cpp
class Solution {
public:
    int maxDistance(vector<int>& position, int m) {
        int n = position.size();
		// sort the given arr
        sort(position.begin(),position.end());
        int ans = -1;
        int l = 0, r = position[n-1];
		
        while(l <= r ){
          int mid = l + (r-l)/2;
          int cnt = 1, last_pos = position[0];
          bool flag = false;
          for(int i = 0; i<n; i++){
            if(position[i] - last_pos >= mid){
              cnt++;
              if(cnt == m){
                flag = true;
                break;
              }
              last_pos = position[i];
            }
          }
          if(flag){
            ans = max(ans,mid);
            l = mid +1;
          }
          else r = mid-1;
        }
      return ans;
    }
};

File: 1552. Magnetic Force Between Two Balls.java
class Solution {
    public boolean check(int mid,int position[],int m){
        int cnt=1;
        int n=position.length;
        int temp=position[0];
        // System.out.println(temp);
        for(int i=0;i<n;i++){
            int diff=Math.abs(position[i]-temp);
            if(diff>=mid){
                temp=position[i];
                //   System.out.println(temp);
                cnt++;
            }
        }
        if(cnt>=m){
            return true;
        }
        return false;
    }
    public int maxDistance(int[] position, int m) {
        int right=-1;
        int n=position.length;
        for(int i=0;i<n;i++){
            right=Math.max(position[i],right);
        }
        int left=0;
        int ans=-1;
        Arrays.sort(position);
        while(left<=right){
            int mid=left+(right-left)/2;
            if(check(mid,position,m)==true){
                ans=mid;
                left=mid+1;
            }
            else{
                right=mid-1;
            }
        }
        return ans;
    }
}
